src/dense.jl

```diff
@@ -1004,13 +1003,14 @@ sqrt(::AbstractMatrix)
 function sqrt(A::AbstractMatrix{T}) where {T<:Union{Real,Complex}}
     if checksquare(A) == 0
         return copy(float(A))
-    elseif isdiag(A) && (T <: Complex || all(x -> x ≥ zero(x), diagview(A)))
-        # Real Diagonal sqrt requires each diagonal element to be positive
-        return applydiagonal(sqrt, A)
+    elseif isdiag(A)
+        if T <: Real && any(<(0), diagview(A))
+            return applydiagonal(sqrt ∘ complex, A)
+        else
+            return applydiagonal(sqrt, A)
+        end
     elseif ishermitian(A)
-        sqrtHermA = sqrt(Hermitian(A))
-        PS = parent(sqrtHermA)
-        return ishermitian(sqrtHermA) ? copytri_maybe_inplace(PS, 'U', true) : PS
+        return parent(sqrt(Hermitian(A)))
```

Comment:
Similar concern about exposing the `parent` as above, as it may not be fully initialized or contain meaningful values.